design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,296 @@
+a push query the response will remain open until the connection is closed or the query is explicitly
+terminated.
+
+#### Terminating queries
+
+Push queries can be explicitly terminated by the client by making a request to this endpoint
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/query-terminate" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added here for the sake of clarity but the real JSON must not contain newlines)
+
+````
+{
+"query-id": "xyz123", <----- the ID of the query to terminate
+}
+
+````
+ 
+
+#### Inserts
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/insert-stream" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added for clarity, the real JSON must not contain newlines):
+
+````
+{
+"stream": "my-stream" <----- The name of the KSQL stream to insert into
+"acks": true          <----- If true then a stream of acks will be returned in the response
+}
+
+````
+
+Followed by zero or more JSON objects representing the values to insert:
+
+````
+{
+"col1" : "val1",
+"col2": 2.3,
+"col3", true
+}
+````
+Each JSON object will be separated by a new line.
+
+To terminate the insert stream the client should end the request.
+
+If acks are requested then the response will be written to the response when each row has been
+successfully committed to the underlying topic. Rows are committed in the order they are provided.
+Each ack in the response is just an empty JSON object, separated by newlines:
+
+````
+{}
+{}
+{}
+{}
+````
+
+#### Errors
+
+Apropriate status codes will be returned from failed requests. The response will also contain JSON
+with further information on the error:
+
+{
+    "error_code": <Error code>
+    "message": <Error Message>
+}
+
+### Migration of existing "REST" API
+
+We will migrate the existing Jetty based "REST" API to the new Vert.x based implementation as-is or with
+minor modifications.
+
+We will migrate the existing Jetty specific plug-ins to Vert.x
+
+### Server implementation
+
+The new server API will be implemented using Vert.x
+
+The implementation will be designed in a reactive / non blocking way in order to provide the best performance / scalability characteristics with
+low resource usage. This will also influence the overall threading model of the server and position us with a better, more scalability internal
+server architecture that will help future proof the ksqlDB server.
+
+## What is not in scope
+
+* We are not creating a new RESTful API.
+* We are not currently looking to write clients for languages other than Java and JavaScript (although those may follow later)
+* We will not initially allow for consuming directly from Kafka topics using the client (although tha may follow later)
```

Comment:
just want to clarify that, if we choose that route, the encoding would still remain the same? i.e Kafka might just replace the http/2 transport and all of this will be transparently hidden under the client?

Reply:
No idea. Let's cross that bridge if we ever come to it. It's certainly not in scope for this KLIP :)

Reply:
Definitely not in scope of this KLIP, but it would be good to look at whether the new client interfaces are built to easily wrap the Kafka clients in a variety of languages. The KSQL Server shouldn't become a Kafka REST proxy. And ideally a KSQL client shouldn't have to implement the Kafka protocol.

Reply:
Whether we decide to support direct access to Kafka topics from the ksqlDB client in the future or not, and how we decide to implement that, e.g. wrapping the Kafka client or proxying through ksqlDB (I would prefer proxying as it would be a pain to have to configure the client to know about two sets of servers) has no impact on the form of the client API that we're proposing here, afaict.